### DISCUSSION TITLE
Level inconsistency when processing chemistry fields from inline data (RRFS-CMAQ)

### DISCUSSION THREAD

**Discussion by edwardstrobach** [dtcenter/METplus#1053] [Configuration] [answered]
@edwardstrobach, please note that I've edited this Discussion by migrating the content over from dtcenter/METplus#1070 here. Is this what you intended? Thanks @JohnHalleyGotway

I'm processing grib output from the RRFS-CMAQ system with new max/ave.
output from post upp. I've made all the adjustments but am facing an issue
that appears linked to how the levels are being specified. The original
message is attached below. In the past I set the forecast level to "L1".
I don't think it works for this case. Notice that the original output was
on sigma levels while this output is on hybrid levels.

Hi Lin,

I made some structural changes in the code to accommodate the differences
in file structure. In the past we separated the fields out and grouped
days 1, 2, and 3 for each file. For instance, for OZMAX8, there will be
three entries in a grib file corresponding to day 1, 2, and 3. In your
example you lump all fields into one file. In the future we may want to
adapt the offline approach to separating data so that we don't have to
change the code.

Unfortunately, the error still occurs. You can see this in my log file (
/gpfs/dell2/emc/modeling/noscrub/Edward.Strobach/metplus_aq/CMAQ/logs/master_metplus.log.20210722174824).
I only applied this to OZMAX1, so please look at that field, not the
others. The references in the log suggest that everything is pointed to
the right locations. For instance, I parsed out the OZMAX1 field from
your post output and put it a ptmp directory that can be found here (
/gpfs/dell2/ptmp/Edward.Strobach/metplus_aqmax/archive/r111_v1/20190815/out_ozmax1.grb2
)

This is what the header info reads for that file; notice the OZMAX1 field
is there:
[Edward.Strobach@v71a3 scripts]$
/gpfs/dell1/nco/ops/nwprod/grib_util.v1.1.1/exec/wgrib2
/gpfs/dell2/ptmp/Edward.Strobach/metplus_aqmax/archive/r111_v1/20190815/out_ozmax1.grb2
1:0:d=2019081512:OZMAX1:1 hybrid level:16-39 hour ave fcst:

The step to generate the observations for verification, using the prepbufr
data and converting that to netcdf file format (that is stored here:
/gpfs/dell2/emc/modeling/noscrub/Edward.Strobach/metplus_aq/CMAQ/aqmmax/aqmax1/r111_v1/prepbufr.aqm.2019081500.nc
http://prepbufr.aqm.2019081500.nc), worked and is supposed to be used
and combined with the out_ozmax1.grb2 file to generate point stat files.
Unfortunately, that doesn't happen because of this error:

WARNING:WARNING: process_fcst_climo_files() -> no fields matching
OZMAX1/L1 found in file:
/gpfs/dell2/ptmp/Edward.Strobach/metplus_aqmax/archive/r111_v1/20190815/out_ozmax1.grb2WARNING:ERROR
:ERROR : process_fcst_climo_files() -> no requested forecast data found!
Exiting...ERROR :07/22 17:48:32.013 metplus.PointStat (command_runner.py:

Note also that I never changed the FCST name or level information. I
shouldn't have to since they are supposedly consistent. Note, however,
that when wgrib2'ing the file that we have used for the offline system,
that I get this:

[Edward.Strobach@v71a1 20190807]$
/gpfs/dell1/nco/ops/nwprod/grib_util.v1.1.1/exec/wgrib2
aqm.t12z.max_1hr_o3.148.grib2 1:0:d=2019080712:OZMAX1:1 sigma
level:-2147483641--2147483618 hour ave fcst:2:144034:d=2019080712:OZMAX1:1
sigma level:17-40 hour ave fcst:

I highlighted in blue the difference in the level information. The
vertical coordinate from the inline system is a hybrid sigma-pressure
coordinate system, I believe, while the offline system is pure sigma. I'm
not sure what the level distinction should be. Could it be that the "L1"
specification is resulting in this error? What then would be the
appropriate specification. Will need help to figure this out I think.
Once this is resolved, then I think it should work and we can then generate
the stats needed. Thanks for working and coordinating on this issue. Let
me know what needs to be done or if more information is needed.

**Comment by JohnHalleyGotway** [comment]
@edwardstrobach, OK, I've logged onto WCOSS to take a look at this issue. As I understand it, your question is really this... How should I set up the MET config file to read the first record from the following grib2 file?

/gpfs/dell2/ptmp/Edward.Strobach/metplus_aqmax/archive/r111_v1/20190815/out_ozmax1.grb2

Unfortunately, I can't find that file. In fact, the /gpfs/dell2/ptmp/Edward.Strobach/metplus_aqmax/archive directory is empty. Please tell me where I can find that file to debug more... or just attach it in a zip file to this discussion.

Looking in the wgrib2 output you sent:
```
[Edward.Strobach@v71a1 20190807]$
/gpfs/dell1/nco/ops/nwprod/grib_util.v1.1.1/exec/wgrib2 aqm.t12z.max_1hr_o3.148.grib2
1:0:d=2019080712:OZMAX1:1 sigma level:-2147483641--2147483618 hour ave fcst:
2:144034:d=2019080712:OZMAX1:1 sigma level:17-40 hour ave fcst:
```

I am most concerned about the forecast hour values listed for the first record: **-2147483641--2147483618 hour ave fcst**

To me that indicates a poorly formatted GRIB2 header.

**Comment by edwardstrobach** [comment]
Hi John,

That is correct.  The impression I get from this exercise is that the issue
seems related to how the level is being specified.  You might find
something different, however.  Yes, those files are stored in ptmp which
gets wiped frequently.  I reran the script and the ptmp directory has been
repopulated, so you should be able to view the contents within that
directory now.

**Reply by edwardstrobach** [reply]
Hello, did you happen to get a chance at looking at the files?  If not, then I'll likely have to rerun again.  Thanks

**Reply by JohnHalleyGotway** [reply]
My apologies Ed, I did not. I was out of the office last week on vacation and dropped the ball. But I'm available to debug today and the remainder of this week.

**Reply by edwardstrobach** [reply]
No problem John.  Everyone needs a vacation.  Let me reproduce the issue and get back to you. Thanks!

**Reply by edwardstrobach** [reply]
Hi John, I reproduced the issue (I hope). The output from the "/gpfs/dell2/emc/modeling/noscrub/Edward.Strobach/metplus_aq/CMAQ/logs/master_metplus.log.20210819170724" has the same error as the original file.

As originally done, I separate the grib file and create grib output for each field.  You can look at referenced files here:
/gpfs/dell2/ptmp/Edward.Strobach/metplus_aqmax/archive/r111_v1/20190815
and 
/gpfs/dell2/ptmp/Edward.Strobach/com/aqm/r111_v1/post_fv3_2019081512/

I've been running a test script from this location: /gpfs/dell2/emc/modeling/save/Edward.Strobach/MET/METplus/scripts/run_metplus_aqmax_test.sh

You can see that the wgrib file output from the forecast file that is used is this:
[Edward.Strobach@m71a1 scripts]$ /gpfs/dell1/nco/ops/nwprod/grib_util.v1.1.1/exec/wgrib2 /gpfs/dell2/ptmp/Edward.Strobach/metplus_aqmax/archive/r111_v1/20190815/out_ozmax1.grb2 
1:0:d=2019081512:OZMAX1:1 hybrid level:16-39 hour ave fcst:

You can also see that the error references that forecast file.  Notice the similarities in name.  This has led me to believe that this could be a "level" issue.  In the original script the level was sigma.  This is hybrid.  Don't know what the specifier should be as a result:

WARNING:
WARNING: MetGrib2DataFile::data_plane_array() -> No matching records found for 'OZMAX1/L1'
WARNING:
DEBUG 2:
DEBUG 2: --------------------------------------------------------------------------------
DEBUG 2:
DEBUG 2: Reading data for OZMAX1/L1.
WARNING:
WARNING: process_fcst_climo_files() -> no fields matching OZMAX1/L1 found in file: /gpfs/dell2/ptmp/Edward.Strobach/metplus_aqmax/archive/r111_v1/20190815/out_ozmax1.grb2
WARNING:
ERROR  :
ERROR  : process_fcst_climo_files() -> no requested forecast data found!  Exiting...
ERROR  :


Note also that this occurs during the pointstat step.

**Reply by JohnHalleyGotway** [reply]
Ed, thanks for posting the data. I pulled down the sample file and was able to plot it using plot_data_plane right away:
```
MET-main_v10.0/met/bin/plot_data_plane 20190815/out_ozmax1.grb2 plot.ps 'name="OZMAX1"; level="L1";' -v 4
```
<img width="428" alt="Screen Shot 2021-08-19 at 12 12 34 PM" src="https://user-images.githubusercontent.com/21087144/130122178-1233f51d-609a-463c-8ae2-d56d9a7f70d6.png">
So I'm confident that MET is able to read data from these files. The most likely problem is some issue in your configuration.

Here's the log file to which you pointed me:
/gpfs/dell2/emc/modeling/noscrub/Edward.Strobach/metplus_aq/CMAQ/logs/master_metplus.log.20210819170724

Scrolling up from the warning/error messages, I see METplus reporting this configuration setting:
```
export FCST_FIELD="{ name=\"OZMAX1\"; level=\"L1\"; valid_time= \"20190816_04\"; }";
```
But that valid time you're requesting (20190816_04) does NOT match the valid time reported by wgrib2 (20190817_03):
```
wgrib2 -V 20190815/out_ozmax1.grb2
1:0:vt=2019081703:1 hybrid level:16-39 hour ave fcst:OZMAX1 Ozone Daily Max from 1-hour Average [ppbV]:
...
```
So that's why MET does not find any data... because you're telling it to read data for a timestamp that is not present in the input file.

Hope that helps clarify.

**Comment by edwardstrobach** [comment]
Okay.  I see.  The inline data does indeed have a shift by one hour
relative to my old set-up.  I did account for those changes elsewhere when
creating the separate grib files.  I guess I can adjust that using
FCST_VAR1_OPTIONS

**Comment by edwardstrobach** [comment]
Thanks.  I'm able to create point stat files and stat files, but they are
not being populated.  There's probably some mismatch still.  I'm surprised
that I had to do a 23 hour shift from the original scripts I used.  The
files not being populated seems related to continued mismatch.  I will
focus on one variable (OZMAX1) and try to resolve that issue.  The steps
indicate success, but there has to be something in the file that points to
no matched pairs and the reason for that.  I do see now that my FCST_FIELD
is correctly specified at least:

08/19 19:32:44.514 metplus.PointStat (command_builder.py:290) DEBUG:
FCST_FIELD={ name="OZMAX1"; level="L1"; valid_time= "20190817_03"; }

**Comment by edwardstrobach** [comment, marked as answer]
FYI,

I believe this is happening because the obs. are coming from 08/15/2019
(i.e.,
/gpfs/dell2/emc/modeling/noscrub/Edward.Strobach/metplus_aq/CMAQ/aqmmax/aqmax1/r111_v1/
prepbufr.aqm.2019081500.nc).  The output from the run is from 08/15/2019.
The wgrib2 command indicates that:

1:0:d=2019081512:OZMAX1:1 hybrid level:16-39 hour ave fcst:

I guess the 08/17/2019 03 is represented by the addition of the two numbers
in blue.  That would result in 20190817_03.  Would you recommend that I
change my valid time to match that for the pb2nc step?  Maybe I'm wrong,
but since it is a 16-39 hour average, it would seem that I would need data
from both 08/16/2019 and 08/17/2019 to successfully get all the matched
pairs I need?  Sorry for all the emails, just trying to understand better.

**Reply by JohnHalleyGotway** [reply]
Ed, just to clarify, running wgrib2 with/without the -V option is useful:
```
wgrib2 out_ozmax1.grb2 
1:0:d=2019081512:OZMAX1:1 hybrid level:16-39 hour ave fcst:
```
Here, "d=2019081512" indicate the model initialization time.
```
wgrib2 -V out_ozmax1.grb2 
1:0:vt=2019081703:1 hybrid level:16-39 hour ave fcst:OZMAX1 Ozone Daily Max from 1-hour Average [ppbV]:
    ndata=18527:undef=0:mean=35.6577:min=11.1044:max=95.3544
...
```
And here, "vt=2019081703" indicates the valid time. Note the "-vt" option also prints the valid time.

As for how to compare the model output to the point observations, its really up to NOAA/EMC folks to decide. I'd recommend that you discuss these details with @PerryShafran-NOAA. Do you want to verify against the raw point observation values? Perhaps compute matched pairs for all observations within the time window? Perhaps only use observation values close to the end of the time window? Or perhaps you want to compute the average obs value at each station over a 24-hour time window that matches the 16-39 hour forecast time window? I believe that all should be possible, but I'd recommend giving it some thought and discussing internally.

**Reply by PerryShafran-NOAA** [reply]
I would not think that we would want to change what we have been doing in terms of this type of verification. I'm trying to follow what has changed since y'all helped set up this type of verification for me previously.

**Comment by edwardstrobach** [comment]
I'm keeping my offline scripts (similar to yours) separate from the inline
scripts.  I need to test inline verification as well.  John mentioned the
different things I could do.  I think computing the average obs value at
each station over a 24-hour time window that matches the 16-39 hour
forecast time window is the way to go.  I'll work on that tomorrow and see
if anything comes from that.  Thanks again John for your help.

**Comment by edwardstrobach** [comment]
Hello,

I'm following up with this discussion with some new findings.  PointStat does run to completion successfully when making the suggested change but there are no matched pairs and thus nothing is being stored in the point stat files.  The log file (**/gpfs/dell2/emc/modeling/noscrub/Edward.Strobach/metplus_aq/CMAQ/logs/master_metplus.log.20210914144647**) shows that the recommended change passed through:

**09/14 14:46:50.688 metplus.PointStat (command_builder.py:290) DEBUG: FCST_FIELD={ name="OZMAX1"; level="L1"; valid_time= "20190817_03"; }**

Originally the valid_time was 20190816_04.  You are correct about the level not being the issue.  The message below also shows that for the whole region there are zero matched pairs:
**DEBUG 2: For OZMAX1/L1 found 1 forecast levels, 0 climatology mean levels, and 0 climatology standard deviation levels.
DEBUG 2:
DEBUG 2: --------------------------------------------------------------------------------
DEBUG 2:
DEBUG 2: Searching 5076 observations from 5076 messages.
DEBUG 2:
DEBUG 2: --------------------------------------------------------------------------------
DEBUG 2:
DEBUG 2: Processing OZMAX1/L1 versus COPO/A1, for observation type AIRNOW_MAX_230000, over region FULL, for interpolation method NEAREST(1), using 0 matched pairs.
DEBUG 3: Number of matched pairs   = 0
DEBUG 3: Observations processed    = 5076
DEBUG 3: Rejected: station id      = 0
DEBUG 3: Rejected: obs type        = 0
DEBUG 3: Rejected: valid time      = 5076
DEBUG 3: Rejected: bad obs value   = 0**

The point stat file is created as mentioned earlier, but it is empty:  

_[Edward.Strobach@m71a3 scripts]$ ls -ltr /gpfs/dell2/emc/modeling/noscrub/Edward.Strobach/metplus_aq/CMAQ/aqmmax/stat/aqmax1/r111_v1//point_stat_R111_V1_AIRNOW_390000L_20190817_030000V.stat
-rw-r--r-- 1 Edward.Strobach emcmodel **240** Sep 14 14:46 /gpfs/dell2/emc/modeling/noscrub/Edward.Strobach/metplus_aq/CMAQ/aqmmax/stat/aqmax1/r111_v1//point_stat_R111_V1_AIRNOW_390000L_20190817_030000V.stat_

It probably has something to do with how the data is written.  The script that I'm adjusting tailored for the offline file structure.  Here is a bit of a comparison showing these differences:

**Offline (the one that works)**

_maximum ozone from 1-hour average_

[Edward.Strobach@m71a1 aqm.**20210910**]$ **/gpfs/dell1/nco/ops/nwprod/grib_util.v1.1.1/exec/wgrib2 -v aqm.t12z.max_1hr_o3.148.grib2**

1:0:vt=**2021091105**:1 sigma level:-**2147483641--2147483618 hour ave** fcst:OZMAX1 Ozone Daily Max from 1-hour Average [ppbV]:
    ndata=117130:undef=0:mean=43.1067:min=12.49:max=82.82
    grid_template=30:winds(grid):
	Lambert Conformal: (442 x 265) input WE:SN output WE:SN res 8
	Lat1 21.821000 Lon1 239.372000 LoV 263.000000
	LatD 33.000000 Latin1 33.000000 Latin2 45.000000
	LatSP 0.000000 LonSP 0.000000
	North Pole (442 x 265) Dx 12000.000000 m Dy 12000.000000 m mode 8

2:141165:vt=**2021091205**:1 sigma level:**17-40 hour ave** fcst:OZMAX1 Ozone Daily Max from 1-hour Average [ppbV]:
    ndata=117130:undef=0:mean=42.8771:min=14.46:max=80.53
    grid_template=30:winds(grid):
	Lambert Conformal: (442 x 265) input WE:SN output WE:SN res 8
	Lat1 21.821000 Lon1 239.372000 LoV 263.000000
	LatD 33.000000 Latin1 33.000000 Latin2 45.000000
	LatSP 0.000000 LonSP 0.000000
	North Pole (442 x 265) Dx 12000.000000 m Dy 12000.000000 m mode 8

3:278591:**vt=2021091305**:1 sigma level:**41-64 hour ave** fcst:OZMAX1 Ozone Daily Max from 1-hour Average [ppbV]:
    ndata=117130:undef=0:mean=42.1417:min=18.13:max=78.51
    grid_template=30:winds(grid):
	Lambert Conformal: (442 x 265) input WE:SN output WE:SN res 8
	Lat1 21.821000 Lon1 239.372000 LoV 263.000000
	LatD 33.000000 Latin1 33.000000 Latin2 45.000000
	LatSP 0.000000 LonSP 0.000000
	North Pole (442 x 265) Dx 12000.000000 m Dy 12000.000000 m mode 8

The first record has a valid time of 2021091105, which marks the ending hour of the -7-16 averaging interval.  The second record has a valid time of 2021091205, which marks the ending hour of the 17-40 averaging interval.  And finally, the third record has a valid time of 2021091305, which marks the ending hour of the 41-64 averaging interval.  These correspond to days 1, 2, and 3.  

Running pointstat generates three of these files corresponding to each day, all of which are populated:

**-rw-r--r-- 1 Edward.Strobach emcmodel   108426 Sep 14 07:23 point_stat_PROD_AIRNOW_160000L_20210912_040000V.stat
-rw-r--r-- 1 Edward.Strobach emcmodel   108785 Sep 14 07:24 point_stat_PROD_AIRNOW_400000L_20210912_040000V.stat
-rw-r--r-- 1 Edward.Strobach emcmodel   108789 Sep 14 07:24 point_stat_PROD_AIRNOW_640000L_20210912_040000V.stat**

**Inline System (the one that doesn't work)**

When I run wgrib2 on the subsetted file for OZMAX1 I see the following:
[Edward.Strobach@m71a1 20190815]$ /gpfs/dell1/nco/ops/nwprod/grib_util.v1.1.1/exec/wgrib2 -V out_ozmax1.grb2 
1:0:vt=**2019081703**:1 hybrid level:**16-39 hour ave** fcst:OZMAX1 Ozone Daily Max from 1-hour Average [ppbV]:
    ndata=18527:undef=0:mean=35.6577:min=11.1044:max=95.3544
    grid_template=30:winds(N/S):
	Lambert Conformal: (191 x 97) input WE:SN output WE:SN res 0
	Lat1 25.116486 Lon1 239.270370 LoV 262.500000
	LatD 38.500000 Latin1 38.500000 Latin2 38.500000
	LatSP 0.000000 LonSP 0.000000
	North Pole (191 x 97) Dx 25000.000000 m Dy 25000.000000 m mode 0

Right now there is only one record, not 3.  However, there is a very important and subtle difference between how the offline is being processed v. how inline is being processed.  

**Offline**
For the Sept. 12 example you see FCST_FIELD being defined as
**FCST_FIELD="{ name=\"OZMAX1\"; level=\"L1\"; valid_time= \"20210912_04\"; }"**
for day 1.  

The corresponding _vt_ record shows the following in the 1-hour max ozone file:
1:0:vt=**2021091305**:1 sigma level:-2147483641--2147483618 hour ave fcst:OZMAX1 Ozone Daily Max from 1-hour Average [ppbV]:
 
Notice it is off by 1 hour.

**Inline**
As mentioned the forecast field is being defined as

**09/14 14:46:50.688 metplus.PointStat (command_builder.py:290) DEBUG: FCST_FIELD={ name="OZMAX1"; level="L1"; valid_time= "20190817_03"; }**
 
while the corresponding _vt_ is 
1:0:vt=**2019081703**:1 hybrid level:**16-39 hour ave** fcst:OZMAX1 Ozone Daily Max from 1-hour Average [ppbV]:

While the inline system have the same definition, the offline system is off by 1 hour.  I tried using valid_time = **20190817_02** based on what I'm seeing for the offline system as a way to test for consistency.  That did not work, however.  I'm unsure how to proceed.

**Comment by edwardstrobach** [comment]
Oh, and I do want to clarify...I don't believe this is an issue with metplus; rather, it is an issue with how the records are being written.  I think there needs to be a 1 hour shift in the record in order for metplus to detect matched pairs.  Do you agree?

**Comment by JohnHalleyGotway** [comment]
Hi Ed, yes, I definitely agree that the issue is in how that averaging period is defined... the 17-40 averaging window certainly differs from 16-39. And if you're explicitly specifying the "valid_time" to be evaluated then data for that time would definitely need to exist to make it all work.

However, I really can't tell you what is correct here. Should it be 17-40 or should it be 16-39? I have no idea. I'd say that's up to you and the data producers to decide.

You have some options for how to handle this via METplus. Obviously, if you can determine what is correct and make the metadata correct, that'd be the best choice. But if you can't make the metadata correct, you could account for the discrepancy in the METplus setup.

As you already know, specifying the valid_time in the configuration string tells MET to only use data for that valid time. And that's a useful way of selecting a record from a GRIB file that includes multiple OZMAX1 records:
``` 
FCST_FIELD="{ name="OZMAX1"; level="L1"; valid_time= "20210912_04"; }"
```
By default, Point-Stat defines the matching time window relative to the valid time of the forecast. For example, a matching time window 1 hour +/- the forecast valid time is given by:
```
obs_window = { beg = -3600; end = 3600; }
```
However, the Point-Stat "-obs_valid_beg" and "-obs_valid_end" command line options can be used to override that default:
```
		"-obs_valid_beg time" in YYYYMMDD[_HH[MMSS]] sets the beginning of the matching time window (optional).
		"-obs_valid_end time" in YYYYMMDD[_HH[MMSS]] sets the end of the matching time window (optional).
```
If you need to account for a 1-hour offset between the forecast valid time and the verifying observations, you could do so either using the "obs_window" config option or explicitly set the "-obs_valid_beg/end" command line options. Either would work. For example, if you want a +/- 1 hour window but for 1 hour earlier, you could just subtract 1 hour (3600 seconds) from each entry:
```
obs_window = { beg = -7200; end = 0; }
```

So again, if you can fix the data to make it more correct, do so. If not, you should be able to account for the discrepancy in your configuration files. I wish I could tell you what the right answer is, but I can't. I imagine that @PerryShafran-NOAA struggled through the same details when he originally set up this verification workflow.

**Reply by edwardstrobach** [reply]
Thanks John for explanations and the discussion of the obs_window feature.  Just to clarify, I can keep the OBS_WINDOW specification for the pb2nc step the same but should specify -7200 for obs_window_beg and 0 for obs_window_end for point stat only to account for the hour shift.

**Reply by JohnHalleyGotway** [reply]
Yes, you certainly can keep using the "obs_window" config option. I was just explaining how it works and how you can set it to account for an offset in time between the forecast and observation data. Exactly how you set it is up to you. I don't know if you shift the values forward one hour or backward one hour, but with some trial and error, it should be pretty obvious.
